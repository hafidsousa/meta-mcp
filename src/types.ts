/**
 * @fileoverview Type definitions for Facebook Marketing API integration
 * 
 * @note For AI Assistants:
 * - All monetary values should be in cents (2000 = $20.00)
 * - Status values should use standard Facebook API values (e.g., 'ACTIVE', 'PAUSED')
 * 
 * @see https://developers.facebook.com/docs/marketing-api/reference/ad-campaign-group
 */

// Remove SDK import and define our own types

export interface FacebookConfig {
  /** Access token with ads_management permission */
  accessToken: string;
  /** Facebook App ID from developer console (optional when using MCP tools) */
  appId?: string;
  /** Facebook App Secret from developer console (optional when using MCP tools) */
  appSecret?: string;
  /** Facebook Ad Account ID (without 'act_' prefix) */
  adAccountId: string;
}

/**
 * Campaign Objectives as specified in Facebook Marketing API
 * @see https://developers.facebook.com/docs/marketing-api/reference/ad-campaign-group
 */
export enum CampaignObjective {
  OUTCOME_TRAFFIC = 'OUTCOME_TRAFFIC',
  OUTCOME_ENGAGEMENT = 'OUTCOME_ENGAGEMENT', 
  OUTCOME_SALES = 'OUTCOME_SALES',
  OUTCOME_LEADS = 'OUTCOME_LEADS',
  OUTCOME_AWARENESS = 'OUTCOME_AWARENESS',
  OUTCOME_APP_PROMOTION = 'OUTCOME_APP_PROMOTION',
  APP_INSTALLS = 'APP_INSTALLS',
  BRAND_AWARENESS = 'BRAND_AWARENESS',
  CONVERSIONS = 'CONVERSIONS',
  EVENT_RESPONSES = 'EVENT_RESPONSES',
  LEAD_GENERATION = 'LEAD_GENERATION',
  LINK_CLICKS = 'LINK_CLICKS',
  LOCAL_AWARENESS = 'LOCAL_AWARENESS',
  MESSAGES = 'MESSAGES',
  PAGE_LIKES = 'PAGE_LIKES',
  POST_ENGAGEMENT = 'POST_ENGAGEMENT',
  PRODUCT_CATALOG_SALES = 'PRODUCT_CATALOG_SALES',
  REACH = 'REACH',
  STORE_VISITS = 'STORE_VISITS',
  VIDEO_VIEWS = 'VIDEO_VIEWS'
}

/**
 * Campaign Status values
 */
export enum CampaignStatus {
  ACTIVE = 'ACTIVE',
  PAUSED = 'PAUSED',
  DELETED = 'DELETED',
  ARCHIVED = 'ARCHIVED'
}

/**
 * Special Ad Categories
 */
export enum SpecialAdCategory {
  EMPLOYMENT = 'EMPLOYMENT',
  HOUSING = 'HOUSING',
  CREDIT = 'CREDIT',
  ISSUES_ELECTIONS_POLITICS = 'ISSUES_ELECTIONS_POLITICS',
  NONE = 'NONE'
}

/**
 * Bid Strategy options
 */
export enum BidStrategy {
  LOWEST_COST_WITHOUT_CAP = 'LOWEST_COST_WITHOUT_CAP',
  LOWEST_COST_WITH_BID_CAP = 'LOWEST_COST_WITH_BID_CAP',
  COST_CAP = 'COST_CAP',
  LOWEST_COST_WITH_MIN_ROAS = 'LOWEST_COST_WITH_MIN_ROAS'
}

export interface AdLabel {
  id: string;
  name: string;
  created_time?: string;
  updated_time?: string;
}

/**
 * Campaign configuration for creating new campaigns
 */
export interface CampaignConfig {
  /** Campaign name following format: [Objective]-[Target]-[Date] */
  name: string;
  /** Campaign objective */
  objective: CampaignObjective;
  /** Initial campaign status */
  status: CampaignStatus;
  /** Special ad category restrictions */
  specialAdCategories?: SpecialAdCategory[];
  /** Campaign spend cap in cents */
  spendCap?: number;
  /** Bid strategy for campaign when using campaign budget optimization */
  bidStrategy?: BidStrategy;
  /** Daily budget in cents */
  dailyBudget?: number;
  /** Lifetime budget in cents */
  lifetimeBudget?: number;
  /** Campaign budget optimization enabled */
  campaignBudgetOptimization?: boolean;
  /** ROAS (Return on Ad Spend) goal value */
  minRoasTargetValue?: number;
  /** Ad Labels associated with this campaign */
  adLabels?: { name: string }[];
  /** When to start running this campaign */
  startTime?: string;
  /** When to stop running this campaign */
  stopTime?: string;
  /** ID of the boosted object */
  boostedObjectId?: string;
  /** When the campaign was created */
  createdTime?: string;
  /** The source of this campaign */
  source_campaign_id?: string;
  /** ID of the Page or app that is being advertised */
  promotedObject?: {
    pageId?: string;
    applicationId?: string;
    objectStoreUrl?: string;
    customEventType?: string;
    offerId?: string;
    pixelId?: string;
    productSetId?: string;
    productCatalogId?: string;
    placePageSetId?: string;
  };
  /** Whether to use default buying type */
  useDefaultBuyingType?: boolean;
  /** Whether this campaign is optimized for buying type */
  isSkadnetworkAttribution?: boolean;
}

export interface AdSetConfig {
  /** Ad Set name following format: [Target Audience]-[Placement]-[Date] */
  name: string;
  /** Parent campaign ID */
  campaignId: string;
  /** Daily budget in cents */
  dailyBudget?: number;
  /** Lifetime budget in cents */
  lifetimeBudget?: number;
  /** When to start running ads (ISO 8601 format with timezone) */
  startTime?: string;
  /** Optional end time for the ad set (ISO 8601 format with timezone) */
  endTime?: string;
  /** Initial ad set status */
  status?: 'ACTIVE' | 'PAUSED';
  /** Optional bid amount in cents */
  bidAmount?: number;
  /** Bid strategy type */
  bidStrategy?: 'LOWEST_COST_WITHOUT_CAP' | 'LOWEST_COST_WITH_BID_CAP' | 'COST_CAP';
  /** Targeting configuration */
  targeting: {
    /** Geographic targeting */
    geoLocations?: {
      countries?: string[];
      regions?: { key: string }[];
      cities?: { key: string, radius?: number, distance_unit?: string }[];
      zips?: { key: string }[];
      geo_markets?: { key: string }[];
      electoral_districts?: { key: string }[];
    };
    /** Minimum age */
    ageMin?: number;
    /** Maximum age */
    ageMax?: number;
    /** Gender targeting (1 = male, 2 = female) */
    genders?: number[];
    /** Interest targeting */
    interests?: { id: string, name?: string }[];
    /** Publisher platforms (facebook, instagram, audience_network, messenger) */
    publisherPlatforms?: string[];
    /** Placement positions (feed, right_hand_column, instant_article, etc.) */
    facebookPositions?: string[];
    /** Instagram positions (stream, story, explore, reels) */
    instagramPositions?: string[];
    /** Device platforms (mobile, desktop) */
    devicePlatforms?: string[];
    /** Behavior targeting */
    behaviors?: { id: string, name?: string }[];
    /** Excluded interests */
    excludedInterests?: { id: string, name?: string }[];
    /** Excluded demographics */
    excludedDemographics?: { id: string, name?: string }[];
    /** Locales (languages) */
    locales?: number[];
    /** User OS */
    userOs?: string[];
    /** Mobile device targeting */
    userDevice?: string[];
    /** Flexible spec for advanced targeting with AND/OR logic */
    flexibleSpec?: Record<string, any>[];
    /** Excluded specs */
    exclusions?: Record<string, any>;
    /** Custom audiences */
    customAudiences?: { id: string, name?: string }[];
  };
  /** Optimization goal */
  optimizationGoal?: 'NONE' | 'APP_INSTALLS' | 'AD_RECALL_LIFT' | 'ENGAGED_USERS' | 
                     'EVENT_RESPONSES' | 'IMPRESSIONS' | 'LEAD_GENERATION' | 'QUALITY_LEAD' | 
                     'LINK_CLICKS' | 'OFFSITE_CONVERSIONS' | 'PAGE_LIKES' | 'POST_ENGAGEMENT' | 
                     'QUALITY_CALL' | 'REACH' | 'LANDING_PAGE_VIEWS' | 'VISIT_INSTAGRAM_PROFILE' | 
                     'VALUE' | 'THRUPLAY' | 'DERIVED_EVENTS' | 'CONVERSIONS';
  /** Billing event type */
  billingEvent?: 'APP_INSTALLS' | 'IMPRESSIONS' | 'LINK_CLICKS' | 'NONE' | 
                 'PAGE_LIKES' | 'POST_ENGAGEMENT' | 'THRUPLAY';
  /** Promoted object for specific campaign types (page_id, application_id, etc.) */
  promotedObject?: {
    pageId?: string;
    applicationId?: string;
    objectStoreUrl?: string;
    customEventType?: string;
    offerId?: string;
    pixelId?: string;
    pixelRule?: string;
    [key: string]: any;
  };
  /** Destination type for ads */
  destinationType?: string;
  /** Pacing type (standard or day_parting) */
  pacingType?: string[];
  /** Facebook attribution spec */
  attributionSpec?: Record<string, any>[];
  /** Use average cost bidding */
  useAverageCost?: boolean;
  /** Time-based ad scheduling */
  adSchedules?: {
    days: number[];
    hours: number[];
    minutes: number[];
  }[];
}

export interface AdCreativeConfig {
  /** Creative name for reference */
  name: string;
  /** Creative ID for existing creatives */
  creative_id?: string;
  /** Ad headline - keep under 40 characters */
  title: string;
  /** Ad body text - keep under 125 characters */
  body: string;
  /** Image URL - must be public accessible */
  imageUrl?: string;
  /** Video URL if using video creative */
  videoUrl?: string;
  /** Landing page URL */
  linkUrl: string;
  /** Call to action button text */
  callToAction?: string;
  /** Additional creative URL parameters */
  urlTags?: string;
  /** Creative format (single_image, carousel, video, etc) */
  format?: string;
  /** Object story spec for Page post ads */
  objectStorySpec?: {
    pageId?: string;
    instagramActorId?: string;
    linkData?: Record<string, any>;
    photoData?: Record<string, any>;
    videoData?: Record<string, any>;
  };
}

export interface AdConfig {
  /** Ad name following format: [Creative Type]-[Target]-[Date] */
  name: string;
  /** Parent ad set ID */
  adsetId: string;
  /** Initial ad status */
  status: 'ACTIVE' | 'PAUSED' | 'DELETED' | 'ARCHIVED';
  /** Creative configuration */
  creative: AdCreativeConfig;
  /** Tracking specs */
  trackingSpecs?: Record<string, any>[];
  /** Optional bid amount in cents */
  bidAmount?: number;
  /** Conversion domain for tracking */
  conversionDomain?: string;
  /** Ad labels associated with this ad */
  adlabels?: AdLabel[];
  /** Display sequence within campaign */
  display_sequence?: number;
  /** Engagement audience flag */
  engagement_audience?: boolean;
}

// Define our own campaign, ad set and ad objects instead of using SDK types
export interface Campaign {
  /** Campaign ID */
  id: string;
  /** Campaign name */
  name: string;
  /** Ad account ID */
  account_id?: string;
  /** Campaign objective */
  objective: string;
  /** Campaign status */
  status: CampaignStatus;
  /** List of ad labels */
  adlabels?: AdLabel[];
  /** Bid strategy */
  bid_strategy?: BidStrategy;
  /** The ID of the boosted object */
  boosted_object_id?: string;
  /** Budget rebalance flag */
  budget_rebalance_flag?: boolean;
  /** Budget remaining */
  budget_remaining?: string;
  /** Buying type (AUCTION or RESERVED) */
  buying_type?: string;
  /** Campaign group active status indicator */
  can_create_brand_lift_study?: boolean;
  /** Can use spend cap flag */
  can_use_spend_cap?: boolean;
  /** Whether the campaign was created to be a conversion campaign */
  configured_status?: CampaignStatus;
  /** Creation time */
  created_time?: string;
  /** Daily budget in cents */
  daily_budget?: string;
  /** Effective status */
  effective_status?: string;
  /** Date campaign is scheduled to end */
  end_time?: string;
  /** Has secondary skadnetwork reporting flag */
  has_secondary_skadnetwork_reporting?: boolean;
  /** Is SKAdNetwork attribution enabled */
  is_skadnetwork_attribution?: boolean;
  /** Issues info */
  issues_info?: any[];
  /** Campaign lifetime budget in cents */
  lifetime_budget?: string;
  /** Campaign optimization goal */
  optimization_goal?: string;
  /** Pacing type for the campaign */
  pacing_type?: string[];
  /** The promoted object */
  promoted_object?: {
    pixel_id?: string;
    page_id?: string;
    custom_event_type?: string;
    application_id?: string;
    object_store_url?: string;
    product_set_id?: string;
    product_catalog_id?: string;
    place_page_set_id?: string;
    offer_id?: string;
  };
  /** List of regular categories for ads */
  smart_promotion_type?: string;
  /** The source campaign ID */
  source_campaign_id?: string;
  /** Special ad categories applied to this campaign */
  special_ad_categories?: SpecialAdCategory[];
  /** Special ad category country */
  special_ad_category_country?: string[];
  /** Campaign spend cap in cents */
  spend_cap?: string;
  /** Campaign start time */
  start_time?: string;
  /** Campaign status */
  stop_time?: string;
  /** Topline ID associated with campaign */
  topline_id?: string;
  /** Last update time */
  updated_time?: string;
  /** Allow additional keys */
  [key: string]: any;
}

export interface AdSet {
  id: string;
  name: string;
  campaign_id: string;
  daily_budget?: string;
  lifetime_budget?: string;
  bid_amount?: string;
  bid_strategy?: string;
  billing_event?: string;
  optimization_goal?: string;
  pacing_type?: string[];
  start_time?: string;
  end_time?: string;
  targeting?: Record<string, any>;
  status: string;
  promoted_object?: Record<string, any>;
  attribution_spec?: Record<string, any>[];
  adlabels?: Record<string, any>[];
  execution_options?: string[];
  destination_type?: string;
  use_average_cost?: boolean;
  ad_schedules?: Record<string, any>[];
  [key: string]: any;
}

export interface Ad {
  id: string;
  name: string;
  adset_id: string;
  campaign_id?: string;
  creative?: Record<string, any>;
  status: string;
  effective_status?: string;
  configured_status?: string;
  tracking_specs?: Record<string, any>[];
  adlabels?: AdLabel[];
  bid_amount?: string;
  conversion_domain?: string;
  createdTime?: string;
  updatedTime?: string;
  [key: string]: any;
}

export interface CampaignResponse {
  /** Created campaign ID */
  id: string;
  /** Whether creation was successful */
  success: boolean;
  /** Campaign data if creation was successful */
  data?: Campaign;
  /** Error details if creation failed */
  error?: any;
}

export interface AdSetResponse {
  /** Created ad set ID */
  id: string;
  /** Whether creation was successful */
  success: boolean;
  /** Ad set data if creation was successful */
  data?: AdSet;
  /** Error details if creation failed */
  error?: any;
}

export interface AdResponse {
  /** Created ad ID */
  id: string;
  /** Whether creation was successful */
  success: boolean;
  /** Ad data if creation was successful */
  data?: Ad;
  /** Error details if creation failed */
  error?: any;
}

export interface FacebookAdAccount {
  id: string;
  name: string;
  account_id: string;
  account_status: number;
  amount_spent: string;
  currency: string;
  [key: string]: any;
} 