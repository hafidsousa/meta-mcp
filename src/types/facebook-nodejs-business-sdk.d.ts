declare module 'facebook-nodejs-business-sdk' {
  export class FacebookAdsApi {
    static init(accessToken: string): FacebookAdsApi;
    call<T>(method: string, path: string, params?: Record<string, any>): Promise<T>;
  }

  export class Campaign {
    constructor(id: string);
    update(data: { status?: string }): Promise<Campaign>;
    id: string;
  }

  export class AdSet {
    constructor(id: string);
    update(data: { status?: string }): Promise<AdSet>;
    id: string;
  }

  export class Ad {
    constructor(id: string);
    update(data: { status?: string }): Promise<Ad>;
    id: string;
  }

  export class AdAccount {
    constructor(id: string);
  }
} 