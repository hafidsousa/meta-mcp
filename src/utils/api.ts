/**
 * @fileoverview API utilities for Facebook Marketing API
 * 
 * @note IMPORTANT: This is a pure pass-through proxy. All requests and responses
 * are passed directly to and from the Facebook API without modification.
 */

import { log } from '../config';

/**
 * Makes a Graph API request
 * @param baseUrl Base URL for the API
 * @param path API endpoint path
 * @param accessToken Facebook access token
 * @param method HTTP method
 * @param params Request parameters
 * @returns Promise with API response
 */
export async function apiRequest(
  baseUrl: string,
  path: string, 
  accessToken: string,
  method: 'GET' | 'POST' | 'DELETE' = 'GET', 
  params: Record<string, any> = {}
): Promise<any> {
  // Always include access token
  params.access_token = accessToken;
  
  const url = `${baseUrl}/${path}`;
  let response;
  
  log(`Making ${method} request to: ${url}`);
  
  if (method === 'GET') {
    // For GET requests, append params to URL
    const queryParams: Record<string, string> = {};
    
    // Convert all values to strings
    Object.entries(params).forEach(([key, value]) => {
      if (typeof value === 'object') {
        queryParams[key] = JSON.stringify(value);
      } else {
        queryParams[key] = String(value);
      }
    });
    
    const queryString = new URLSearchParams(queryParams).toString();
    log(`GET params: ${queryString}`);
    response = await fetch(`${url}?${queryString}`);
  } else {
    // For POST/DELETE, use request body
    const body = new URLSearchParams();
    
    for (const [key, value] of Object.entries(params)) {
      if (typeof value === 'object') {
        body.append(key, JSON.stringify(value));
      } else {
        body.append(key, String(value));
      }
    }
    
    log(`${method} params: ${JSON.stringify(Object.fromEntries(body.entries()))}`);
    
    response = await fetch(url, {
      method,
      body,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
    });
  }
  
  // Treat data as any to maintain the pass-through nature
  const data = await response.json() as any;
  
  if (!response.ok || data.error) {
    log(`API Error response: ${JSON.stringify(data)}`);
    
    // Extract error information for better error messages
    const errorMessage = data.error?.message || 
                         data.error_description ||
                         data.error_msg ||
                         data.error_message || 
                         `API request failed with status ${response.status}`;
    
    // Create an error with the Facebook API error message
    const error = new Error(errorMessage);
    
    // Attach the full error response for better debugging
    (error as any).fbErrorResponse = data.error || data;
    (error as any).httpStatus = response.status;
    
    throw error;
  }
  
  return data;
} 