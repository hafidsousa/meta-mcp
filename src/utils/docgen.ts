/**
 * @fileoverview Documentation generation utilities
 * These utilities generate tool documentation from TypeScript types
 */

import { Tool } from "@modelcontextprotocol/sdk/types.js";
import * as SharedTypes from '../types/shared';

/**
 * Generates a Tool definition for MCP from a TypeScript type
 * This ensures tool documentation stays in sync with actual types
 * 
 * @param name Tool name
 * @param description Tool description
 * @param configType Name of the configuration or entity type
 * @param additionalParams Additional parameters for the tool
 * @returns Tool definition
 */
export function generateToolFromType(
  name: string,
  description: string,
  configType: string,
  additionalParams: Record<string, any> = {}
): Tool {
  const inputSchema: any = {
    type: "object",
    properties: {
      ...additionalParams
    },
    required: []
  };
  
  // Add config parameter if it's a config type
  if (configType.includes('Config')) {
    inputSchema.properties.config = {
      type: "object",
      description: `${configType.replace('Config', '')} configuration object`
    };
    inputSchema.required.push('config');
  }
  
  // Add parameters for entity IDs
  if (name.includes('pause') || name.includes('update') || name.includes('get')) {
    if (name.includes('Campaign')) {
      inputSchema.properties.campaignId = {
        type: "string",
        description: "ID of the campaign"
      };
      inputSchema.required.push('campaignId');
    } else if (name.includes('AdSet')) {
      inputSchema.properties.adSetId = {
        type: "string",
        description: "ID of the ad set"
      };
      inputSchema.required.push('adSetId');
    } else if (name.includes('Ad') && !name.includes('AdSet')) {
      inputSchema.properties.adId = {
        type: "string",
        description: "ID of the ad"
      };
      inputSchema.required.push('adId');
    }
  }
  
  return {
    name,
    description,
    inputSchema
  };
}

/**
 * Generates a comprehensive set of tool definitions
 * @returns Array of Tool definitions
 */
export function generateAllToolDefinitions(): Tool[] {
  return [
    // Campaign tools
    generateToolFromType(
      "createCampaign",
      "Creates a new Facebook ad campaign. This is the first step in the ad creation process. A campaign defines the overall objective and budget strategy.",
      "CampaignConfig"
    ),
    
    generateToolFromType(
      "getCampaigns",
      "Gets all campaigns for the ad account. Returns an array of campaign objects with their IDs, names, status, and other details.",
      "Campaign",
      {
        limit: {
          type: "number",
          description: "Maximum number of campaigns to return (default: 25, max: 100)"
        },
        status: {
          type: "string",
          description: "Filter campaigns by status",
          enum: Object.values(SharedTypes.CampaignStatus)
        }
      }
    ),
    
    generateToolFromType(
      "getCampaign",
      "Gets details for a specific campaign by ID.",
      "Campaign"
    ),
    
    generateToolFromType(
      "pauseCampaign",
      "Pauses an active campaign, which also pauses all associated ad sets and ads.",
      "Campaign"
    ),
    
    generateToolFromType(
      "updateCampaign",
      "Updates an existing campaign with new configuration values.",
      "CampaignConfig"
    ),
    
    // Ad Set tools
    generateToolFromType(
      "createAdSet",
      "Creates a new ad set within a campaign. Ad sets define targeting, budget, scheduling, and bidding for your ads.",
      "AdSetConfig"
    ),
    
    generateToolFromType(
      "getAdSets",
      "Gets all ad sets for a specific campaign.",
      "AdSet",
      {
        campaignId: {
          type: "string",
          description: "ID of the campaign"
        }
      }
    ),
    
    generateToolFromType(
      "getAdSet",
      "Gets details for a specific ad set by ID.",
      "AdSet"
    ),
    
    generateToolFromType(
      "getAccountAdSets",
      "Gets all ad sets for the configured ad account.",
      "AdSet",
      {
        limit: {
          type: "number",
          description: "Maximum number of ad sets to return"
        },
        status: {
          type: "string",
          description: "Filter ad sets by status",
          enum: Object.values(SharedTypes.CampaignStatus)
        }
      }
    ),
    
    generateToolFromType(
      "pauseAdSet",
      "Pauses an active ad set and all its ads.",
      "AdSet"
    ),
    
    generateToolFromType(
      "updateAdSet",
      "Updates an existing ad set with new configuration values.",
      "AdSetConfig"
    ),
    
    // Ad tools
    generateToolFromType(
      "createAd",
      "Creates a new ad within an ad set. Ads are the actual content shown to users.",
      "AdConfig"
    ),
    
    generateToolFromType(
      "getAds",
      "Gets all ads for a specific ad set.",
      "Ad",
      {
        adSetId: {
          type: "string",
          description: "ID of the ad set"
        }
      }
    ),
    
    generateToolFromType(
      "getAd",
      "Gets details for a specific ad by ID.",
      "Ad"
    ),
    
    generateToolFromType(
      "getAccountAds",
      "Gets all ads for the configured ad account.",
      "Ad",
      {
        limit: {
          type: "number",
          description: "Maximum number of ads to return"
        },
        status: {
          type: "string",
          description: "Filter ads by status",
          enum: Object.values(SharedTypes.CampaignStatus)
        }
      }
    ),
    
    generateToolFromType(
      "pauseAd",
      "Pauses an active ad.",
      "Ad"
    ),
    
    generateToolFromType(
      "updateAd",
      "Updates an existing ad with new configuration values.",
      "AdConfig"
    ),
    
    // Account tools
    generateToolFromType(
      "getAvailableAdAccounts",
      "Gets all available ad accounts for the current user or page.",
      "FacebookConfig",
      {}
    )
  ];
} 