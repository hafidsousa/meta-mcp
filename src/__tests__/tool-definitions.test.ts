/**
 * @fileoverview Tests for tool definitions
 * Uses snapshot testing to ensure tool definitions don't change unintentionally
 */

import { generateAllToolDefinitions } from '../utils/docgen';
import { Tool } from '@modelcontextprotocol/sdk/types.js';

describe('Tool Definitions', () => {
  it('should match the snapshot for all tool definitions', () => {
    const tools = generateAllToolDefinitions();
    
    // This will create a snapshot on first run, and compare on subsequent runs
    expect(tools).toMatchSnapshot();
  });
  
  it('should include all required Facebook Marketing API tools', () => {
    const tools = generateAllToolDefinitions();
    const toolNames = tools.map((tool: Tool) => tool.name);
    
    // Core Facebook Marketing API operations
    const requiredToolNames = [
      'createCampaign',
      'getCampaigns',
      'pauseCampaign',
      'updateCampaign',
      'createAdSet',
      'getAdSets',
      'pauseAdSet',
      'updateAdSet',
      'createAd',
      'getAds',
      'pauseAd',
      'updateAd',
      'getAvailableAdAccounts'
    ];
    
    requiredToolNames.forEach((toolName: string) => {
      expect(toolNames).toContain(toolName);
    });
  });
  
  it('should have proper descriptions for all tools', () => {
    const tools = generateAllToolDefinitions();
    
    tools.forEach((tool: Tool) => {
      // Ensure description is not empty
      expect(tool.description?.trim()).not.toBe('');
      
      // Ensure description is reasonable length
      expect(tool.description?.length).toBeGreaterThan(10);
      
      // Ensure input schema properties have descriptions
      if (tool.inputSchema?.type === 'object' && tool.inputSchema.properties) {
        Object.values(tool.inputSchema.properties).forEach((prop: any) => {
          if (typeof prop === 'object' && prop.description) {
            expect(prop.description.trim()).not.toBe('');
          }
        });
      }
    });
  });
}); 