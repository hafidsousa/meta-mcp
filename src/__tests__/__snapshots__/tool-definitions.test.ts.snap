// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tool Definitions should match the snapshot for all tool definitions 1`] = `
[
  {
    "description": "Creates a new Facebook ad campaign. This is the first step in the ad creation process. A campaign defines the overall objective and budget strategy.",
    "inputSchema": {
      "properties": {
        "config": {
          "description": "Campaign configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
      ],
      "type": "object",
    },
    "name": "createCampaign",
  },
  {
    "description": "Gets all campaigns for the ad account. Returns an array of campaign objects with their IDs, names, status, and other details.",
    "inputSchema": {
      "properties": {
        "campaignId": {
          "description": "ID of the campaign",
          "type": "string",
        },
        "limit": {
          "description": "Maximum number of campaigns to return (default: 25, max: 100)",
          "type": "number",
        },
        "status": {
          "description": "Filter campaigns by status",
          "enum": [
            "ACTIVE",
            "PAUSED",
            "DELETED",
            "ARCHIVED",
          ],
          "type": "string",
        },
      },
      "required": [
        "campaignId",
      ],
      "type": "object",
    },
    "name": "getCampaigns",
  },
  {
    "description": "Gets details for a specific campaign by ID.",
    "inputSchema": {
      "properties": {
        "campaignId": {
          "description": "ID of the campaign",
          "type": "string",
        },
      },
      "required": [
        "campaignId",
      ],
      "type": "object",
    },
    "name": "getCampaign",
  },
  {
    "description": "Pauses an active campaign, which also pauses all associated ad sets and ads.",
    "inputSchema": {
      "properties": {
        "campaignId": {
          "description": "ID of the campaign",
          "type": "string",
        },
      },
      "required": [
        "campaignId",
      ],
      "type": "object",
    },
    "name": "pauseCampaign",
  },
  {
    "description": "Updates an existing campaign with new configuration values.",
    "inputSchema": {
      "properties": {
        "campaignId": {
          "description": "ID of the campaign",
          "type": "string",
        },
        "config": {
          "description": "Campaign configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
        "campaignId",
      ],
      "type": "object",
    },
    "name": "updateCampaign",
  },
  {
    "description": "Creates a new ad set within a campaign. Ad sets define targeting, budget, scheduling, and bidding for your ads.",
    "inputSchema": {
      "properties": {
        "config": {
          "description": "AdSet configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
      ],
      "type": "object",
    },
    "name": "createAdSet",
  },
  {
    "description": "Gets all ad sets for a specific campaign.",
    "inputSchema": {
      "properties": {
        "adSetId": {
          "description": "ID of the ad set",
          "type": "string",
        },
        "campaignId": {
          "description": "ID of the campaign",
          "type": "string",
        },
      },
      "required": [
        "adSetId",
      ],
      "type": "object",
    },
    "name": "getAdSets",
  },
  {
    "description": "Gets details for a specific ad set by ID.",
    "inputSchema": {
      "properties": {
        "adSetId": {
          "description": "ID of the ad set",
          "type": "string",
        },
      },
      "required": [
        "adSetId",
      ],
      "type": "object",
    },
    "name": "getAdSet",
  },
  {
    "description": "Gets all ad sets for the configured ad account.",
    "inputSchema": {
      "properties": {
        "adSetId": {
          "description": "ID of the ad set",
          "type": "string",
        },
        "limit": {
          "description": "Maximum number of ad sets to return",
          "type": "number",
        },
        "status": {
          "description": "Filter ad sets by status",
          "enum": [
            "ACTIVE",
            "PAUSED",
            "DELETED",
            "ARCHIVED",
          ],
          "type": "string",
        },
      },
      "required": [
        "adSetId",
      ],
      "type": "object",
    },
    "name": "getAccountAdSets",
  },
  {
    "description": "Pauses an active ad set and all its ads.",
    "inputSchema": {
      "properties": {
        "adSetId": {
          "description": "ID of the ad set",
          "type": "string",
        },
      },
      "required": [
        "adSetId",
      ],
      "type": "object",
    },
    "name": "pauseAdSet",
  },
  {
    "description": "Updates an existing ad set with new configuration values.",
    "inputSchema": {
      "properties": {
        "adSetId": {
          "description": "ID of the ad set",
          "type": "string",
        },
        "config": {
          "description": "AdSet configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
        "adSetId",
      ],
      "type": "object",
    },
    "name": "updateAdSet",
  },
  {
    "description": "Creates a new ad within an ad set. Ads are the actual content shown to users.",
    "inputSchema": {
      "properties": {
        "config": {
          "description": "Ad configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
      ],
      "type": "object",
    },
    "name": "createAd",
  },
  {
    "description": "Gets all ads for a specific ad set.",
    "inputSchema": {
      "properties": {
        "adId": {
          "description": "ID of the ad",
          "type": "string",
        },
        "adSetId": {
          "description": "ID of the ad set",
          "type": "string",
        },
      },
      "required": [
        "adId",
      ],
      "type": "object",
    },
    "name": "getAds",
  },
  {
    "description": "Gets details for a specific ad by ID.",
    "inputSchema": {
      "properties": {
        "adId": {
          "description": "ID of the ad",
          "type": "string",
        },
      },
      "required": [
        "adId",
      ],
      "type": "object",
    },
    "name": "getAd",
  },
  {
    "description": "Gets all ads for the configured ad account.",
    "inputSchema": {
      "properties": {
        "adId": {
          "description": "ID of the ad",
          "type": "string",
        },
        "limit": {
          "description": "Maximum number of ads to return",
          "type": "number",
        },
        "status": {
          "description": "Filter ads by status",
          "enum": [
            "ACTIVE",
            "PAUSED",
            "DELETED",
            "ARCHIVED",
          ],
          "type": "string",
        },
      },
      "required": [
        "adId",
      ],
      "type": "object",
    },
    "name": "getAccountAds",
  },
  {
    "description": "Pauses an active ad.",
    "inputSchema": {
      "properties": {
        "adId": {
          "description": "ID of the ad",
          "type": "string",
        },
      },
      "required": [
        "adId",
      ],
      "type": "object",
    },
    "name": "pauseAd",
  },
  {
    "description": "Updates an existing ad with new configuration values.",
    "inputSchema": {
      "properties": {
        "adId": {
          "description": "ID of the ad",
          "type": "string",
        },
        "config": {
          "description": "Ad configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
        "adId",
      ],
      "type": "object",
    },
    "name": "updateAd",
  },
  {
    "description": "Gets all available ad accounts for the current user or page.",
    "inputSchema": {
      "properties": {
        "adId": {
          "description": "ID of the ad",
          "type": "string",
        },
        "config": {
          "description": "Facebook configuration object",
          "type": "object",
        },
      },
      "required": [
        "config",
        "adId",
      ],
      "type": "object",
    },
    "name": "getAvailableAdAccounts",
  },
]
`;
